src/visualization/plot_recipes.jl

```diff
@@ -709,6 +516,115 @@ RecipesBase.@recipe function f(u, semi::AbstractSemidiscretization;
+@recipe function f(u, semi::DGMultiSemidiscretizationHyperbolic;
+                   solution_variables=cons2cons, grid_lines=true)
+  return PlotData2D(u, semi)
+end
+
+# If `u` is an `Array{<:SVectors}` and not a `StructArray`, convert it to a `StructArray` first.
+function PlotData2D(u::Array{<:SVector, 2}, mesh::AbstractMeshData, equations, dg::DGMulti, cache;
+                    solution_variables=nothing, nvisnodes=2*polydeg(dg))
+  nvars = length(first(u))
+  u_structarray = StructArray{eltype(u)}(ntuple(_->zeros(eltype(first(u)), size(u)), nvars))
+  for (i, u_i) in enumerate(u)
+    u_structarray[i] = u_i
+  end
+
+  # re-dispatch to PlotData2D with mesh, equations, dg, cache arguments
+  return PlotData2D(u_structarray, mesh, equations, dg, cache;
+                    solution_variables=solution_variables, nvisnodes=nvisnodes)
+end
+
+# constructor which returns an `UnstructuredPlotData2D` object.
+function PlotData2D(u::StructArray, mesh::AbstractMeshData, equations, dg::DGMulti, cache;
+                    solution_variables=nothing, nvisnodes=2*polydeg(dg))
+
+  rd = dg.basis
+  md = mesh.md
+
+  # Vp = the interpolation matrix from nodal points to plotting points
+  @unpack Vp = rd
+  interpolate_to_plotting_points!(out, x) = mul!(out, Vp, x)
+
+  solution_variables_ = digest_solution_variables(equations, solution_variables)
+  variable_names = SVector(varnames(solution_variables_, equations))
+
+  num_plotting_points = size(Vp, 1)
+  nvars = nvariables(equations)
+  uEltype = eltype(first(u))
+  u_plot_local = StructArray{SVector{nvars, uEltype}}(ntuple(_->zeros(uEltype, num_plotting_points), nvars))
+  u_plot = zeros(SVector{nvars, uEltype}, num_plotting_points, md.num_elements)
+  for e in eachelement(mesh, dg, cache)
+
+    # interpolate solution to plotting nodes element-by-element
+    StructArrays.foreachfield(interpolate_to_plotting_points!, u_plot_local, view(u, :, e))
+
+    # transform nodal values of the solution according to `solution_variables`
+    for (i, u_i) in enumerate(u_plot_local)
+      u_plot[i, e] = solution_variables_(u_i, equations)
+    end
+  end
+
+  # interpolate nodal coordinates to plotting points
+  x_plot, y_plot = map(x->Vp * x, md.xyz) # md.xyz is a tuple of arrays containing nodal coordinates
+
+  # construct a triangulation of the reference plotting nodes
+  t = reference_plotting_triangulation(rd.rstp) # rd.rstp = reference coordinates of plotting points
+
+  xfplot, yfplot = mesh_plotting_wireframe(rd, md, num_plotting_points=nvisnodes)
+
+  # Set the plotting values of solution on faces to nothing - they're not used for Plots.jl since
+  # only 2D heatmap plots are supported through TriplotBase/TriplotRecipes.
+  ufplot = nothing
+
+  return UnstructuredPlotData2D(x_plot, y_plot, u_plot, t, xfplot, yfplot, ufplot, variable_names)
+end
+
+# Series recipe for UnstructuredPlotData2D
+@recipe function f(pds::PlotDataSeries2D{<:UnstructuredPlotData2D})
```

Comment:
Types (except tuples) are not covariant:
```julia
julia> Vector{Float64} <: Vector{Real}
false

julia> Vector{Float64} <: Vector{<:Real}
true
```

Reply:
~~Yes, but `UnstructuredPlotData2D` is a concrete type and so is `PlotDataSeries2D{UnstructuredPlotData2D}`. Or am I missing something?~~

Ah, is it because `UnstructuredPlotData2D` is actually an incomplete type and thus the need for `<:`?

Reply:
It has type parameters. Thus, `UnstructuredPlotData2D` without specific parameters is a `UnionAll`.

Reply:
OK, thanks! In this case, @jlchan please use @ranocha 's, suggestion.